### PR TITLE
New auton commands

### PR DESCRIPTION
**Pull request by vs2961**

### CODE REVIEW

File: src/main/java/frc/robot/commands/auton/DrivetrainRotatePIDCommand.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        System.out.println(getAngle());
+    }
+
+    private boolean onTarget() {
+        return Math.abs(getAngle() - targetAngle) <= 2;
+    }
+
+    protected abstract double getAngle();
+
+    private class GyroPIDSource implements PIDSource {
```
Comment (PrathamRawat) [inline src/main/java/frc/robot/commands/auton/DrivetrainRotatePIDCommand.java]: The class does appear to be used in other places, all the more reason to make it an external `public class`

File: src/main/java/frc/robot/commands/auton/DrivetrainStraightRampingCommand.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    @Override
+    protected void initialize() {
+        super.initialize();
+        Robot.drivetrain.setRamp(SmartDashboard.getNumber("DriveStraight RampSeconds", 2.5));
+        speedPIDController = new PIDController(0, 0, 0, new GyroPIDSource(), new GyroPIDOutput());
+        speedPIDController.setSetpoint(distance);
+        speedPIDController.setAbsoluteTolerance(3);
+        speedPIDController.setPID(SmartDashboard.getNumber("DriveStraightRampingPID P", 0),
```
Comment (PrathamRawat) [inline src/main/java/frc/robot/commands/auton/DrivetrainStraightRampingCommand.java]: Local Variables please

File: src/main/java/frc/robot/commands/auton/DrivetrainDriveStraightCommand.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    protected void end() {
+        super.end();
+        rotationPIDController.setPID(0, 0, 0);
+        rotationPIDController.disable();
+    }
+
+    protected double getGyroPIDOutput() {
+        return gyroPIDOutput;
+    }
+    private class GyroPIDSource implements PIDSource {
```
[the remark this answers is not in the document]
Comment (vs2961) [reply src/main/java/frc/robot/commands/auton/DrivetrainDriveStraightCommand.java]: Other PID commands can simply call the getGyroPIDOutput method.

File: src/main/java/frc/robot/commands/auton/DrivetrainRotatePIDCommand.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+import edu.wpi.first.wpilibj.PIDController;
+import edu.wpi.first.wpilibj.PIDOutput;
+import edu.wpi.first.wpilibj.PIDSource;
+import edu.wpi.first.wpilibj.PIDSourceType;
+import edu.wpi.first.wpilibj.Timer;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import frc.robot.Robot;
+
+public abstract class DrivetrainRotatePIDCommand extends DrivetrainRelativeRotateCommand {
```
[the remark this answers is not in the document]
Comment (vs2961) [reply src/main/java/frc/robot/commands/auton/DrivetrainRotatePIDCommand.java]: That's another file

File: src/main/java/frc/robot/commands/auton/DrivetrainRotatePIDCommand.java (hunk 1 above)
Comment (vs2961) [reply src/main/java/frc/robot/commands/auton/DrivetrainRotatePIDCommand.java]: It's being used for 2 different purposes, though

File: src/main/java/frc/robot/commands/auton/DrivetrainAbsoluteRotateCommand.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+    @Override
+    protected void initialize() {
+        Robot.drivetrain.resetGyro();
+    }
+
+    @Override
+    protected void execute() {
+        if (Robot.drivetrain.getAbsoluteGyroAngle() == targetAngle) {
+            Robot.drivetrain.stop();
```
Comment (bryanlai00) [inline src/main/java/frc/robot/commands/auton/DrivetrainAbsoluteRotateCommand.java]: Robot.drivetrain.stop() is also called in end(), which runs the command twice.

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 6]
```diff
@@ -182,4 +188,26 @@ public void lowGearShift() {
     public void toggleGearShift(){
         gearShift.set(!(gearShift.get()));
     }
+
+    public void setRamp(double rampSeconds) {
```
Comment (bryanlai00) [inline src/main/java/frc/robot/subsystems/Drivetrain.java]: We should only call two motors due to the motors following each other.

**Pull request closed without merging**